### ISSUE TITLE
Issue with parameters "labels"?

### ISSUE THREAD

**Issue by JulienPascal** [floswald/MomentOpt.jl#20]
Hi,

With Edoardo Ciscato, we noticed something strange. It seems like the library "exchanges" the parameters to be estimated at some point (at least their names).

Here is an example, where I want to estimate the mean of 4-dimensional Gaussian process (I fix the variance-covariance matrix to be the identity matrix):
```julia
using MomentOpt
using GLM
using DataStructures
using DataFrames
using Plots
#plotlyjs()
pyplot()

#------------------------------------------------
# Options
#-------------------------------------------------
# Boolean: do you want to save the plots to disk?
savePlots = true

# initialize the problem:
# 4 means
#------------------------
# initial values:
#----------------
pb = OrderedDict("p1" => [0.2,-3,3] , "p2" => [-0.2,-2,2], "p3" => [0.1,0,10], "p4" => [-0.1,-10,0] )
# moments to be matched:
#-----------------------
moms = DataFrame(name=["mu1","mu2","mu3", "mu4"],value=[-1.0,1.0, 5.0, -4.0], weight=ones(4))



"""
    objfunc_normal(ev::Eval)

    GMM objective function to be minized.
    It returns a weigthed distance between empirical and simulated moments

    copy-paste of the function objfunc_norm(ev::Eval)
    I only made minor modifications to the original fuction

"""

function objfunc_normal(ev::Eval)

    start(ev)


    # extract parameters from ev:
    #----------------------------
    mu  = collect(values(ev.params))

    # compute simulated moments
    #--------------------------
    # Monte-Carlo:
    #-------------
    ns = 10000 #number of i.i.d draws from N([a,b], sigma)
    #initialize a multivariate normal N([a,b], sigma)
    #using a = mu[1], b=mu[2]
    sigma = [1.0 ;1.0; 1.0; 1.0]
    randMultiNormal = MomentOpt.MvNormal(mu,MomentOpt.PDiagMat(sigma))
    simM            = mean(rand(randMultiNormal,ns),2) #mean of simulated data
    simMoments = Dict(:mu1 => simM[1], :mu2 => simM[2], :mu3 => simM[3], :mu4 => simM[4])#store simulated moments in a dictionary


    # Calculate the weighted distance between empirical moments
    # and simulated ones:
    #-----------------------------------------------------------
    v = Dict{Symbol,Float64}()
    for (k, mom) in dataMomentd(ev)
        # If weight for moment k exists:
        #-------------------------------
        if haskey(MomentOpt.dataMomentWd(ev), k)
            # divide by weight associated to moment k:
            #----------------------------------------
            v[k] = ((simMoments[k] .- mom) ./ MomentOpt.dataMomentW(ev,k)) .^2
        else
            v[k] = ((simMoments[k] .- mom) ) .^2
        end
    end

    # Set value of the objective function:
    #------------------------------------
    setValue(ev, mean(collect(values(v))))

    # also return the moments
    #-----------------------
    setMoment(ev, simMoments)

    # flag for success:
    #-------------------
    ev.status = 1

    # finish and return
    finish(ev)

    return ev
end



# Initialize an empty MProb() object:
#------------------------------------
mprob = MProb()

# Add structural parameters to MProb():
# specify starting values and support
#--------------------------------------
addSampledParam!(mprob,pb)

# Add moments to be matched to MProb():
#--------------------------------------
addMoment!(mprob,moms)

# Attach an objective function to MProb():
#----------------------------------------
addEvalFunc!(mprob, objfunc_normal)


# estimation options:
#--------------------
# number of iterations for each chain
niter = 1000
# number of chains
# nchains = nprocs()
nchains = 4

opts = Dict("N"=>nchains,
        "maxiter"=>niter,
        "maxtemp"=> 5,
        "coverage"=>0.025,
        "sigma_update_steps"=>10,
        "sigma_adjust_by"=>0.01,
        "smpl_iters"=>1000,
        "parallel"=>true,
        "maxdists"=>[0.05 for i in 1:nchains],
        "mixprob"=>0.3,
        "acc_tuner"=>12.0,
        "animate"=>false)


# plot slices of objective function
#---------------------------------
s = doSlices(mprob,10)

# plot objective function over param values:
#-------------------------------------------
p1 = MomentOpt.plot(s,:value)

if savePlots == true
    Plots.savefig(p1, joinpath(pwd(),"slices_Normal1.svg"))
end


#---------------------------------------
# Let's set-up and run the optimization
#---------------------------------------
# set-up BGP algorithm:
MA = MAlgoBGP(mprob,opts)

# run the estimation:
@time MomentOpt.runMOpt!(MA)

# show a summary of the optimization:
@show MomentOpt.summary(MA)

# Plot histograms for the first chain:
#------------------------------------
p4 = histogram(MA.chains[1])

if savePlots == true
    savefig(p4, joinpath(pwd(),"histogram_chain1.svg"))
end

# Plot the "history" of one chain:
#--------------------------------
p10 = plot(MA.chains[1])
if savePlots == true
    savefig(p10, joinpath(pwd(),"history_chain_1.svg"))
end

```
The algorithm recovers the true parameters ([-1.0,1.0, 5.0, -4.0]) but not in the right order. It finds p1=5.0, p2 = -1, p3 = -4, p4 = 1.0, while it should be p1=-1.0, p2=1.0, p3=5.0, p4 = -4.0.
![histogram_chain1](https://user-images.githubusercontent.com/15772983/40915813-be50097a-67fd-11e8-89db-7f6b5210ac9c.png)
![figure_1](https://user-images.githubusercontent.com/15772983/40915814-bf7db856-67fd-11e8-9fd8-61575a2e05b1.png)

Is it an issue with the plotting functions, or something more serious? 

------------------------------------------------------------------------------------------
Edit: on the other hand, slicing works fine
![slice](https://user-images.githubusercontent.com/15772983/40916434-035c60ca-6800-11e8-8a95-ecb01d30c203.png)

**Comment by floswald** [comment]
oh! this is very odd, yes. no ideas from the top of my head. just: i can see from the plots that you are not using the standard GR backend. (you are running the latest vesion of MomentOpt?). can you check if the `gr()` backend does the same thing?

**Comment by JulienPascal** [comment]
A minor change seems to fix this issue. In the file AlgoBGP.jl
Instead of using a dictionary in this function
```julia
function proposal(c::BGPChain)

    if c.iter==1
        return c.m.initial_value
    else
        ev_old = getLastAccepted(c)
        mu  = paramd(ev_old) # dict of params
        lb = [v[:lb] for (k,v) in c.m.params_to_sample]
        ub = [v[:ub] for (k,v) in c.m.params_to_sample]

        # Transition Kernel is q(.|theta(t-1)) ~ TruncatedN(theta(t-1), Sigma,lb,ub)
        newp = Dict(zip(collect(keys(mu)),mysample(MvNormal(collect(values(mu)),c.sigma),lb,ub,c.smpl_iters)))
        # @debug(logger,"iteration $(c.iter)")
        # @debug(logger,"old param: $(ev_old.params)")
        # @debug(logger,"new param: $newp")

        # flat kernel: random choice in each dimension.
        # newp = Dict(zip(collect(keys(mu)),rand(length(lb)) .* (ub .- lb)))

        return newp
    end

end
```
It is better to use an ordered dictionary:
```julia
function proposal(c::BGPChain)

    if c.iter==1
        return c.m.initial_value
    else
        ev_old = getLastAccepted(c)
        mu  = paramd(ev_old) # dict of params
        lb = [v[:lb] for (k,v) in c.m.params_to_sample]
        ub = [v[:ub] for (k,v) in c.m.params_to_sample]

        # Transition Kernel is q(.|theta(t-1)) ~ TruncatedN(theta(t-1), Sigma,lb,ub)
        newp = OrderedDict(zip(collect(keys(mu)),mysample(MvNormal(collect(values(mu)),c.sigma),lb,ub,c.smpl_iters)))
        # @debug(logger,"iteration $(c.iter)")
        # @debug(logger,"old param: $(ev_old.params)")
        # @debug(logger,"new param: $newp")

        # flat kernel: random choice in each dimension.
        # newp = Dict(zip(collect(keys(mu)),rand(length(lb)) .* (ub .- lb)))

        return newp
    end

end
```

I have to give full credit to @edoardociscato, who suggested this change. 

The order seems to matter, and using a "classical" dictionary will change this order. 

I run the same code with the modification above, and here are the results:
![histogram_chain1_corrected](https://user-images.githubusercontent.com/15772983/40917558-febed486-6803-11e8-8288-a3ffea6a1528.png)
![history_chain1_corrected](https://user-images.githubusercontent.com/15772983/40917566-02d6f788-6804-11e8-898a-30804ca9b882.png)

I can make a pull request with this (small) change.

**Comment by edoardociscato** [comment]
I confirm that I experienced similar problems. Slices also works fine with me: as long the objective function is well-written as in Julien's example, there is no problem.
In the example above, Julien builds an object `pb` of type `OrderedDict` with (correctly labeled) initial values, lower and upper boundaries, and feeds it to `MProb()` with `addSampledParam!`.
My understanding is that the problem arises within the function `next_eval` cited below. We are instructed to feed initial values and support through `OrderedDict`, and, consistently, each evaluation `eval` on each chain has its parameters stored in field `param` of type `OrderedDict`. However, the function `proposal` returns an object `Dict`, with a mixed-up order of parameters. This incorrect order is passed on to new iterations generating several problems. In particular, apart from labeling, parameters are associated with wrong support and wrong variances when using the function `proposal`.
I am checking how to fix this, but I guess the easiest way is to always use the type `OrderedDict` when feeding parameters to a function and returning them.

https://github.com/floswald/MomentOpt.jl/blob/0c90a769d42aad3c29d1f5aecf5a49cd1e63f406/src/mopt/AlgoBGP.jl#L215-L236

**Comment by floswald** [comment]
Good work guys! Yes exactly that should be an ordered dict for precisely
the reason you say.

This library is becoming useable! :-)

**Comment by floswald** [comment]
Pull request very welcome!

**Comment by floswald** [comment]
closed by #21 . can I use this example on the readme? looks great!

**Comment by JulienPascal** [comment]
The code for the estimation of the mean of the multivariate normal? 
Yes of course. It is based on the code I found here: https://github.com/floswald/MomentOpt.jl/blob/master/src/mopt/Examples.jl

**Comment by floswald** [comment]
of course it is. i'm dumb. :+1:

**Comment by JulienPascal** [comment]
A refresh of the readme would be needed though. Because of the issue we just fixed, it shows an estimated value (1.0 -1.0) while it should be (-1.0 1.0)
![capture du 2018-06-04 17-14-09](https://user-images.githubusercontent.com/15772983/40926097-ad33c41a-681b-11e8-9dcf-7bd69b161a63.png)
I can update the readme if you think it is a good idea.
